File: CleverOnion-backend/src/main/java/com/cleveronion/blog/domain/common/aggregate/AggregateRoot.java
package com.cleveronion.blog.domain.common.aggregate;

import com.cleveronion.blog.domain.common.event.DomainEvent;
import com.fasterxml.jackson.annotation.JsonIgnore;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 聚合根基类
 * 提供领域事件管理功能
 */
public abstract class AggregateRoot {
    
    @JsonIgnore
    private final List<DomainEvent> domainEvents = new ArrayList<>();
    
    /**
     * 添加领域事件
     * 
     * @param event 要添加的领域事件
     */
    protected void addDomainEvent(DomainEvent event) {
        if (event != null) {
            this.domainEvents.add(event);
        }
    }
    
    /**
     * 获取所有领域事件（只读）
     * 
     * @return 领域事件列表的只读副本
     */
    @JsonIgnore
    public List<DomainEvent> getDomainEvents() {
        return Collections.unmodifiableList(domainEvents);
    }
    
    /**
     * 清除所有领域事件
     * 通常在事件发布后调用
     */
    public void clearDomainEvents() {
        this.domainEvents.clear();
    }
    
    /**
     * 检查是否有待发布的领域事件
     * 
     * @return true如果有待发布的事件
     */
    @JsonIgnore
    public boolean hasDomainEvents() {
        return !domainEvents.isEmpty();
    }
    
    /**
     * 获取聚合根标识
     * 子类必须实现此方法
     * 
     * @return 聚合根的唯一标识
     */
    public abstract String getAggregateId();
}

File: CleverOnion-blog-k/src/main/kotlin/com/cleveronion/domain/shared/aggregate/AggregateRoot.kt
package com.cleveronion.domain.shared.aggregate

import com.cleveronion.domain.shared.event.DomainEvent

/**
 * 聚合根抽象基类
 * 
 * 聚合根是聚合的唯一入口点，负责维护聚合内部的一致性和业务规则。
 * 聚合根管理领域事件的产生和发布。
 * 
 * @param T 聚合根标识符的类型
 */
abstract class AggregateRoot<T> {
    private val domainEvents = mutableListOf<DomainEvent>()
    
    /**
     * 获取聚合产生的领域事件
     * 
     * @return 领域事件列表的只读副本
     */
    fun getDomainEvents(): List<DomainEvent> = domainEvents.toList()
    
    /**
     * 清除已处理的领域事件
     * 
     * 通常在事件发布后调用此方法清理事件列表
     */
    fun clearDomainEvents() {
        domainEvents.clear()
    }
    
    /**
     * 添加领域事件
     * 
     * 当聚合内发生重要业务事件时，调用此方法记录事件
     * 
     * @param event 要添加的领域事件
     */
    protected fun addDomainEvent(event: DomainEvent) {
        domainEvents.add(event)
    }
    
    /**
     * 聚合根必须实现业务规则验证
     * 
     * 此方法用于验证聚合的业务不变性，确保聚合始终处于有效状态
     */
    abstract fun ensureBusinessRules()
    
    /**
     * 获取聚合根标识符
     * 
     * @return 聚合根的唯一标识符
     */
    abstract fun getId(): T
    
    /**
     * 聚合根相等性比较基于标识符
     * 
     * 两个聚合根如果具有相同的标识符，则认为它们是相等的
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AggregateRoot<*>) return false
        return getId() == other.getId()
    }
    
    /**
     * 基于标识符计算哈希码
     */
    override fun hashCode(): Int {
        return getId()?.hashCode() ?: 0
    }
    
    /**
     * 字符串表示形式
     */
    override fun toString(): String {
        return "${this::class.simpleName}(id=${getId()})"
    }
}